### PR TITLE
BIP 174: Partially Signed Bitcoin Transaction (PSBT) Format

### PR DESCRIPTION
**Pull request by achow101**
This proposal was [submitted to the mailing list](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2017-August/014838.html) on August 18th and has received no relevant commentary, so I am now requesting a BIP number.

### CODE REVIEW

**Comment by achow101** [comment]
@luke-jr number please?

**Comment by dabura667** [comment]
Actually I sent a proposal to the mailing list last year.... the only person who contacted me was @dcousens the head maintainer of bitcoinjs-lib

**Comment by dcousens** [comment]
Maybe we should each hop on a call to talk about a communal approach to this?

**Comment by luke-jr** [comment]
Please use BIP number 174.

File: bip-psbt.mediawiki [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+| Redeem Script<ref>'''Why are redeem scripts and witness scripts global''' Redeem
+  scripts and witness scripts are global data to avoid duplication. Instead of specifying
+  a redeems script and witness script multiple times in inputs that need those scripts,
+  they are specified once in the global data.</ref>
+| The hash160 of the redeem script
+| A redeem script that will be needed to sign a Pay-To-Script-Hash input.
+|-
+| <tt>0x02</tt>
+| Witness Script
+| The sha256 hash of the witness script
```
Comment (luke-jr) [inline bip-psbt.mediawiki]: Why not have the key data include the witness program version, so it can be non-V0?

Comment (achow101) [reply bip-psbt.mediawiki]: This isn't defining witness programs. It is defining the very specific witness script of a P2WSH output.

Comment (luke-jr) [reply bip-psbt.mediawiki]: I don't know what you mean. The value is the specific script, but the key is the hash of it...

**Comment by dabura667** [comment]
https://appear.in/psbt-discussion

I went ahead and claimed this room, and will get notifications when someone enters via the app or via browser (no sign up required)

Be careful though, as the default setting has the webcam automatically turn on, so click the camera icon to disable your camera as soon as you enter.

**Comment by dabura667** [comment]
https://docs.google.com/spreadsheets/d/1PSnBDtu7KCzeEltXySTtFFDaXIoFk7MF7umAxooLRho/edit?usp=sharing

Here's the google doc @dcousens and I were using in January when we discussed this.

@achow101 if you make a comment on one of the cells, I'll add your account as an editor.

**Comment by dabura667** [comment]
https://gist.github.com/dcousens/8e72ff7ddf1f189cb7bbe42d0aeca9ce

Also this was something @dcousens worked on prior.

Perhaps we can find a time when Japan, Australia, and wherever @achow101 lives can meet up for a call?

**Comment by dabura667** [comment]
I will be in SF for Scaling Bitcoin btw.

**Comment by dabura667** [comment]
Here are some comments I made to @dcousens 

I kept thinking down the rabbit hole, and imagined an entire protocol where a wallet can just create the sighash preimage, generate the hash, generate the signature, and reencode the signature into the data so that another party could sign or complete the serialized complete transaction data.

> After thinking for a while, I decided to put my goal into a single sentence.
> 
> "By parsing the data alone, a compliant wallet will know exactly what they need to sign, where to put the signatures, and how to serialize the completed transaction in the event of completion without ant prior agreed upon template."
> 
> By doing so, it would be possible for a wallet that is single sig only to just implement the standard, and be able to sign a multisig without even knowing it is a part of a multisig. It just follows the rules, creates the tx for signing and signs, then adds the signature binary data to the data chunk (URI or otherwise)
> 
> After thinking a lot, there should also be perhaps two separate related BIPs on relay methods and initialization methods.
> 
> ie. Which methods are supported? If we recommend a JSON API, what are the necessary endpoints? Do we support IP direct connection or not? etc. Should there be voluntary discovery like Electrum nodes? Etc etc
> 
> ie. How do you set up the configuration and share it between wallets? How do you "backup" the configurations (having such freedom means that private keys are not our only worry anymore, we must remember which private key was a member of what type of wallet and where is the metadata stored if any (a la Copay stores in BWS)
> 
> I think we could make these three BIPs encompass a large possibility space to allow for true wallet interoperability rather than just have it be a way to encode the same old cookie cutter transactions by using previous knowledge of a popular template etc.

**Comment by achow101** [comment]
> Perhaps we can find a time when Japan, Australia, and wherever @achow101 lives can meet up for a call?

That may be hard to do. I live on the East coast of the US.

File: bip-psbt.mediawiki (hunk 1 above)
Comment (achow101) [reply bip-psbt.mediawiki]: The key is used as a unique identifier for the script. It is not defining anything for the witness program. For future witness programs, I don't think it is safe to assume that future witness programs will also be using the hash of something in the witness program. I also don't think that the types should be overloaded. The key here is specifically one type: a sha256 hash. However to do what you suggest means that the key can have more than one type in the future.

**Comment by dcousens** [comment]
How about 2000 EDT tomorrow? (1000 AEST, 0900 JST)

**Comment by achow101** [comment]
> How about 2000 EDT tomorrow? (1000 AEST, 0900 JST)

OK

**Comment by junderw** [comment]
Sounds good to me. Just hop in the appear.in and we should be good

File: bip-psbt.mediawiki (hunk 1 above)
Comment (sipa) [reply bip-psbt.mediawiki]: I wonder if instead of just witness scripts, a field for generic "SHA256 preimages" would be useful.

**Comment by dabura667** [comment]
https://appear.in/psbt-discussion

@dcousens We're in the chat now.

**Comment by achow101** [comment]
Some TODOs before merging:

[] Sighash type for inputs
[] Include more information about adding new types (appendix of new types)
[] Include number of inputs for future extensibilty with more blocks of data
[] Examples/basic format/pipe format for each key-value pair

**Comment by achow101** [comment]
@junderw @dcousens I have made the changes that we talked about in the call. Does this look good to you guys?

File: bip-0174.mediawiki [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+  </pre>
+|-
+| <tt>0x03</tt>
+| BIP 32 Derivation path, public key, and Master Key fingerprint
+| The public key
+| The master key fingerprint concatenated with the derivation path of the public key. The
+derivation path is represented as 32 bit unsigned integer indexes concatenated
+with each other. This must omit the index of the master key.
+| <pre>
+    <key length>|0x03|<public key>|<value length>|<master key fingerprint>|<32-bit int>|...|<32-bit int>
```
Comment (dcousens) [inline bip-0174.mediawiki]: `value length` being the number of 32-bit int indexes?

File: bip-0174.mediawiki [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+  </pre>
+|-
+| <tt>0x04</tt>
+| Number of inputs
+| None. The key must only contain the 1 byte type.
+| A compact size unsigned integer representing the number of inputs that this transaction has
+| <pre>
+    <0x01>|0x04|<number of inputs>
```
Comment (dcousens) [inline bip-0174.mediawiki]: what is the difference between <0x01> and 0x01,  you fluctuate between both styles?

File: bip-0174.mediawiki (hunk 2 above)
Comment (achow101) [reply bip-0174.mediawiki]: The value length is the length in bytes of the concatenation of `<master key fingerprint>|<32-bit int>|...|<32-bit int>`. The next commit should make this clearer.

File: bip-0174.mediawiki [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+|-
+| <tt>0x03</tt>
+| Sighash Type
+| None. The key must only contain the 1 byte type.
+| The 32-bit unsigned integer recommending a sighash type to be used for this input.
+The sighash type is only a recommendation and the signer does not need to use
+the sighash type specified.
+| <pre>
+    0x01|0x03|<sighash type>
+  </pre>
```
Comment (dcousens) [inline bip-0174.mediawiki]: to clarify,   to omit 50/51 inputs, leaving the 51st input to be signed,  you'd have to include 50x `0x00` separators?
Could there by an optional `vin`?

File: bip-0174.mediawiki (hunk 3 above)
Comment (achow101) [reply bip-0174.mediawiki]: Fixed.

File: bip-0174.mediawiki (hunk 2 above)
Comment (dcousens) [reply bip-0174.mediawiki]: @achow101 isn't `master key fingerprint` fixed size anyway?

Comment (achow101) [reply bip-0174.mediawiki]: > this part of the `<key length> <key> <value length> <value>` approach?

^That

File: bip-0174.mediawiki [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+| <pre>
+    <key length(0x01)>|0x00|<tx length>|<transaction>
+  </pre>
+|-
+| <tt>0x01</tt>
+| Witness UTXO
+| None. The key must only contain the 1 byte type.
+| The entire transaction output in network serialization which the current input spends from.
+| <pre>
+    <key length(0x01)>|0x01|<hash160>|<redeem script length>|<redeem script>
```
Comment (dcousens) [inline bip-0174.mediawiki]: you say TXO, but you have redeem script (and no satoshis/amount?)

File: bip-0174.mediawiki (hunk 4 above)
Comment (achow101) [reply bip-0174.mediawiki]: Added

File: bip-0174.mediawiki (hunk 5 above)
Comment (achow101) [reply bip-0174.mediawiki]: Oops. bad copy paste.

File: bip-0174.mediawiki [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+| <pre>
+    <key length>|0x01|<hash160>|<redeem script length>|<redeem script>
+  </pre>
+|-
+| <tt>0x02</tt>
+| Witness Script
+| The sha256 hash of the witness script
+| A witness script that will be needed to sign a Pay-To-Witness-Script-Hash input.
+| <pre>
+    <key length>|0x02|<sha256>|<witness script length>|<witness script>
```
Comment (dcousens) [inline bip-0174.mediawiki]: maybe `<key length>` should include the type prefix,  then you can simply define `<key length>` as a the length of the `<key>`, and interpreters interpret the first byte of a key as a type.
That way there is less back-and-forth over the definition,  and its consistent.
I suspect this is how your implementation would work internally anyway.

For example, as your definition:
```
<key length> | <key(0x02 | sha256(<witness script>)) | <value length> | <value(witness script)>
```

Hell,  then you could simplify the entire defintion to simply `key` and `value`, with a line at the top here saying `key` and `length` are both prefixed by their `length` as a varint (or whatever).

Comment (dcousens) [reply bip-0174.mediawiki]: If you want to move past this now,  I could submit this as a PR if you'd like.

Comment (achow101) [reply bip-0174.mediawiki]: That is what I originally meant. The latest changes should make that clearer.

**Comment by dcousens** [comment]
> ^ Why are outputs' redeem scripts and witness scripts included? Redeem scripts and witness scripts spent to by an output in this transaction are included so that the user can verify that the transaction they are signing is creating the correct outputs that have the correct redeem and witness scripts. This is best used when the PSBT creator is not trusted by the signers.

:heart: 

ACK

**Pull request merged**